# ML for CyberSecurity

Name: Jash Rathod

Net ID: jsr10000

# Downloading Relevant Files

We will start by downloading the [CSAW-HackML-2020 GitHub Repo](https://github.com/csaw-hackml/CSAW-HackML-2020) and the [data files](https://drive.google.com/drive/folders/1Rs68uH8Xqa4j6UxG53wzD0uyI8347dSq?usp=sharing)

In [1]:
!git clone https://github.com/csaw-hackml/CSAW-HackML-2020.git

Cloning into 'CSAW-HackML-2020'...
remote: Enumerating objects: 220, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 220 (delta 37), reused 17 (delta 17), pack-reused 178
Receiving objects: 100% (220/220), 83.78 MiB | 26.80 MiB/s, done.
Resolving deltas: 100% (83/83), done.


In [2]:
# All data: https://drive.google.com/drive/folders/1Rs68uH8Xqa4j6UxG53wzD0uyI8347dSq?usp=sharing

# bd_test.h5 - https://drive.google.com/file/d/1kxNACo0qFo8QdZgtGHvaA67p4h4RcNIy/view?usp=sharing
# bd_valid.h5 - https://drive.google.com/file/d/1DRKofqVdn2ioh44M45eYZHl_XAW9r3v4/view?usp=sharing
# test.h5 - https://drive.google.com/file/d/1HpahIi-RcvtaRoly_TbuoBzWUaAjVDgt/view?usp=sharing
# valid.h5 - https://drive.google.com/file/d/1nbB5tyUVClSaFvvg3hrFW4wOUj3GtNTf/view?usp=sharing

import gdown

!mkdir -p /content/CSAW-HackML-2020/lab3/data/bd /content/CSAW-HackML-2020/lab3/data/cl

gid_outpath_list = [("1kxNACo0qFo8QdZgtGHvaA67p4h4RcNIy", "/content/CSAW-HackML-2020/lab3/data/bd/bd_test.h5"),
                    ("1DRKofqVdn2ioh44M45eYZHl_XAW9r3v4", "/content/CSAW-HackML-2020/lab3/data/bd/bd_valid.h5"),
                    ("1HpahIi-RcvtaRoly_TbuoBzWUaAjVDgt", "/content/CSAW-HackML-2020/lab3/data/cl/test.h5"),
                    ("1nbB5tyUVClSaFvvg3hrFW4wOUj3GtNTf", "/content/CSAW-HackML-2020/lab3/data/cl/valid.h5")]

for x in gid_outpath_list:
    gdown.download(f"https://drive.google.com/uc?id={x[0]}", x[1])

Downloading...
From: https://drive.google.com/uc?id=1kxNACo0qFo8QdZgtGHvaA67p4h4RcNIy
To: /content/CSAW-HackML-2020/lab3/data/bd/bd_test.h5
100%|██████████| 398M/398M [00:02<00:00, 160MB/s]
Downloading...
From: https://drive.google.com/uc?id=1DRKofqVdn2ioh44M45eYZHl_XAW9r3v4
To: /content/CSAW-HackML-2020/lab3/data/bd/bd_valid.h5
100%|██████████| 716M/716M [00:06<00:00, 119MB/s]
Downloading...
From: https://drive.google.com/uc?id=1HpahIi-RcvtaRoly_TbuoBzWUaAjVDgt
To: /content/CSAW-HackML-2020/lab3/data/cl/test.h5
100%|██████████| 398M/398M [00:03<00:00, 124MB/s]
Downloading...
From: https://drive.google.com/uc?id=1nbB5tyUVClSaFvvg3hrFW4wOUj3GtNTf
To: /content/CSAW-HackML-2020/lab3/data/cl/valid.h5
100%|██████████| 716M/716M [00:04<00:00, 157MB/s]


In [3]:
!du -sh /content/CSAW-HackML-2020/lab3/models/bd_net.h5
!du -sh /content/CSAW-HackML-2020/lab3/models/bd_weights.h5

7.0M	/content/CSAW-HackML-2020/lab3/models/bd_net.h5
2.4M	/content/CSAW-HackML-2020/lab3/models/bd_weights.h5


# Implementation

We will now perform the following steps:

1. Import libraries
2. Load the dataset
3. Define the model class
4. Define function for pruning defense


In [4]:
# Import necessary libraries
import h5py
import numpy as np
import pandas as pd
from tqdm import tqdm

from tensorflow import keras
from tensorflow.keras.models import Model
import tensorflow.keras.backend as K

In [5]:
# File paths for the datasets and the model
neural_network_model_path = "/content/CSAW-HackML-2020/lab3/models/bd_net.h5"
neural_network_model_weights = "/content/CSAW-HackML-2020/lab3/models/bd_weights.h5"

poisoned_dataset_path = "/content/CSAW-HackML-2020/lab3/data/bd/bd_valid.h5"
poisoned_test_path = "/content/CSAW-HackML-2020/lab3/data/bd/bd_test.h5"

clean_dataset_path = "/content/CSAW-HackML-2020/lab3/data/cl/valid.h5"
test_dataset_path = "/content/CSAW-HackML-2020/lab3/data/cl/test.h5"

In [ ]:
# Function to load data from a given file path
def load_dataset(file_path):
    # Open the file and read data
    file_data = h5py.File(file_path, 'r')
    images, labels = np.array(file_data['data']), np.array(file_data['label'])

    # Adjust the dimensions of the images
    images = images.transpose((0, 2, 3, 1))
    return images, labels

# Loading different datasets
clean_images_valid, clean_labels_valid = load_dataset(clean_dataset_path)
clean_images_test, clean_labels_test = load_dataset(test_dataset_path)
poisoned_images_valid, poisoned_labels_valid = load_dataset(poisoned_dataset_path)

In [ ]:
# Function to extract intermediate layer output from a model
def extract_intermediate_layer(model, layer_index):
    # Create a model to output the intermediate layer
    return keras.models.Model(inputs=model.input, outputs=model.layers[layer_index].output)

In [ ]:
# Class definition for the modified model
# This model outputs a repaired model "G"
# `call` takes argument `input_data` (which is "Dvalid"​, a validation dataset of clean, labelled images)
class ModifiedModel(keras.Model):
    def __init__(self, OriginalModel, PrunedModel):
        super(ModifiedModel, self).__init__()
        self.OriginalModel = OriginalModel
        self.PrunedModel = PrunedModel

    def call(self, input_data):
        original_predictions = np.argmax(self.OriginalModel(input_data), axis=1)
        pruned_predictions = np.argmax(self.PrunedModel(input_data), axis=1)
        combined_result = np.array([original_predictions[i] if original_predictions[i] == pruned_predictions[i] else 1283 for i in range(original_predictions.shape[0])])
        result_matrix = np.zeros((original_predictions.shape[0], 1284))
        result_matrix[np.arange(combined_result.size), combined_result] = 1
        return result_matrix

We will start at layer `6` for pruning (`0`-indexed)

In [ ]:
# Function to apply pruning defense mechanism
def apply_pruning_defense(threshold):
    # Specify the layer index for pruning
    target_layer_index = 6
    OriginalModel = keras.models.load_model(neural_network_model_path) # Model "B"
    PrunedModel = keras.models.load_model(neural_network_model_path)

    # Calculate and display original accuracy
    clean_labels_predicted = np.argmax(OriginalModel.predict(clean_images_valid), axis=1)
    original_accuracy = np.mean(np.equal(clean_labels_predicted, clean_labels_valid)) * 100
    print("Original accuracy on clean data: ", original_accuracy)

    # Calculate and display Attack Success Rate (ASR) of the poisoned model
    poisoned_labels_predicted = np.argmax(OriginalModel.predict(poisoned_images_valid), axis=1)
    attack_success_rate = np.mean(np.equal(poisoned_labels_predicted, poisoned_labels_valid)) * 100
    print("Attack Success Rate (ASR) of BadNet: ", attack_success_rate)

    # Get activations from the intermediate layer
    intermediate_layer_output = extract_intermediate_layer(OriginalModel, target_layer_index)
    activations = intermediate_layer_output.predict(clean_images_valid)

    # Compute average activations and start pruning
    average_activations = activations.mean(axis=(0, 1, 2))
    activation_indices_sorted = np.argsort(average_activations)

    # Iteratively prune channels
    for channel in tqdm(activation_indices_sorted):
        # Access and zero-out the target channel in the last layer
        target_layer = PrunedModel.layers[target_layer_index - 1]
        K.set_value(target_layer.kernel[:, :, :, channel], np.zeros_like(target_layer.kernel[:, :, :, channel]))
        K.set_value(target_layer.bias[channel], np.zeros_like(target_layer.bias[channel]))

        # Recalculate accuracy after each pruning step
        # clean_labels_predicted = np.argmax(PrunedModel.predict(clean_images_valid), axis=1)
        # pruned_accuracy = np.mean(np.equal(clean_labels_predicted, clean_labels_valid)) * 100
        pruned_accuracy = np.mean(np.equal(np.argmax(PrunedModel.predict(clean_images_valid), axis=1), clean_labels_valid)) * 100
        print('\nVal accuracy (after pruning channel', channel, '):', pruned_accuracy)

        # Stop pruning if the accuracy drop exceeds the threshold
        if abs(original_accuracy - pruned_accuracy) > threshold:
            break

    # Create an instance of the ModifiedModel class
    refined_model = ModifiedModel(OriginalModel, PrunedModel)

    # Evaluate the refined model on clean test data
    # clean_test_predictions = refined_model(clean_images_test)
    clean_test_accuracy = np.mean(np.equal(np.argmax(refined_model(clean_images_test), axis=1), clean_labels_test)) * 100
    print("Clean test accuracy of Refined Model: ", clean_test_accuracy)

    # Evaluate the refined model on poisoned validation data
    # poisoned_validation_predictions = refined_model(poisoned_images_valid)
    poisoned_validation_accuracy = np.mean(np.equal(np.argmax(refined_model(poisoned_images_valid), axis=1), poisoned_labels_valid)) * 100
    print("ASR of Refined Model: ", poisoned_validation_accuracy) # Attack success rate

    return PrunedModel

## IMPORTANT NOTE

The next 3 code cells produce the repaired models for different threshold values (2%, 4%, 10%). As these demand large System RAM, not all 3 cells can be run in succession. Hence, run only 1 cell for which you need the results. Then restart the runtime, run all the `Implementation` cells and run the other cell (and then the third).

Below, I have produced results for the 3 repaired models, where each run shares the validation, test accuracies and attack success rates.

In [ ]:
# Prune the model with a threshold of 2 and save the repaired model
repaired_model_threshold_2 = apply_pruning_defense(2)
repaired_model_threshold_2.save("model_repaired_threshold_2.h5")
del repaired_model_threshold_2

361/361 [==============================] - 3s 3ms/step
Original accuracy on clean data:  98.64899974019225
361/361 [==============================] - 1s 2ms/step
Attack Success Rate (ASR) of BadNet:  100.0
361/361 [==============================] - 1s 2ms/step


  0%|          | 0/60 [00:00<?, ?it/s]

361/361 [==============================] - 1s 2ms/step


  2%|▏         | 1/60 [00:02<02:27,  2.51s/it]


Val accuracy (after pruning channel 0 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  3%|▎         | 2/60 [00:04<02:14,  2.31s/it]


Val accuracy (after pruning channel 26 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  5%|▌         | 3/60 [00:06<02:05,  2.20s/it]


Val accuracy (after pruning channel 27 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  7%|▋         | 4/60 [00:09<02:07,  2.27s/it]


Val accuracy (after pruning channel 30 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


  8%|▊         | 5/60 [00:11<02:13,  2.43s/it]


Val accuracy (after pruning channel 31 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 10%|█         | 6/60 [00:14<02:10,  2.41s/it]


Val accuracy (after pruning channel 33 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 12%|█▏        | 7/60 [00:16<02:05,  2.37s/it]


Val accuracy (after pruning channel 34 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step

Val accuracy (after pruning channel 36 

 13%|█▎        | 8/60 [00:18<01:59,  2.30s/it]

): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 15%|█▌        | 9/60 [00:20<01:54,  2.25s/it]


Val accuracy (after pruning channel 37 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 17%|█▋        | 10/60 [00:24<02:15,  2.72s/it]


Val accuracy (after pruning channel 38 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 18%|█▊        | 11/60 [00:26<02:07,  2.61s/it]


Val accuracy (after pruning channel 25 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 20%|██        | 12/60 [00:29<02:02,  2.54s/it]


Val accuracy (after pruning channel 39 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 22%|██▏       | 13/60 [00:31<01:55,  2.47s/it]


Val accuracy (after pruning channel 41 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 23%|██▎       | 14/60 [00:33<01:52,  2.44s/it]


Val accuracy (after pruning channel 44 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 25%|██▌       | 15/60 [00:36<01:55,  2.56s/it]


Val accuracy (after pruning channel 45 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 27%|██▋       | 16/60 [00:39<01:50,  2.51s/it]


Val accuracy (after pruning channel 47 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 28%|██▊       | 17/60 [00:41<01:46,  2.47s/it]


Val accuracy (after pruning channel 48 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 30%|███       | 18/60 [00:43<01:42,  2.44s/it]


Val accuracy (after pruning channel 49 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 32%|███▏      | 19/60 [00:46<01:39,  2.44s/it]


Val accuracy (after pruning channel 50 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 33%|███▎      | 20/60 [00:49<01:41,  2.54s/it]


Val accuracy (after pruning channel 53 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 35%|███▌      | 21/60 [00:51<01:37,  2.49s/it]


Val accuracy (after pruning channel 55 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 37%|███▋      | 22/60 [00:53<01:33,  2.45s/it]


Val accuracy (after pruning channel 40 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 38%|███▊      | 23/60 [00:56<01:29,  2.43s/it]


Val accuracy (after pruning channel 24 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 40%|████      | 24/60 [00:58<01:25,  2.39s/it]


Val accuracy (after pruning channel 59 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 42%|████▏     | 25/60 [01:01<01:27,  2.51s/it]


Val accuracy (after pruning channel 9 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 43%|████▎     | 26/60 [01:03<01:25,  2.50s/it]


Val accuracy (after pruning channel 2 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 45%|████▌     | 27/60 [01:06<01:18,  2.39s/it]


Val accuracy (after pruning channel 12 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 47%|████▋     | 28/60 [01:08<01:14,  2.31s/it]


Val accuracy (after pruning channel 13 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 48%|████▊     | 29/60 [01:10<01:12,  2.33s/it]


Val accuracy (after pruning channel 17 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 50%|█████     | 30/60 [01:13<01:13,  2.45s/it]


Val accuracy (after pruning channel 14 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 52%|█████▏    | 31/60 [01:15<01:10,  2.43s/it]


Val accuracy (after pruning channel 15 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 53%|█████▎    | 32/60 [01:17<01:07,  2.42s/it]


Val accuracy (after pruning channel 23 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 55%|█████▌    | 33/60 [01:20<01:04,  2.41s/it]


Val accuracy (after pruning channel 6 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 57%|█████▋    | 34/60 [01:22<01:00,  2.35s/it]


Val accuracy (after pruning channel 51 ): 98.64033948211657
361/361 [==============================] - 1s 3ms/step


 58%|█████▊    | 35/60 [01:25<01:00,  2.43s/it]


Val accuracy (after pruning channel 32 ): 98.64033948211657
361/361 [==============================] - 1s 2ms/step

Val accuracy (after pruning channel 22 ): 

 60%|██████    | 36/60 [01:27<00:56,  2.34s/it]

98.63167922404088
361/361 [==============================] - 1s 2ms/step


 62%|██████▏   | 37/60 [01:29<00:53,  2.32s/it]


Val accuracy (after pruning channel 21 ): 98.65765999826795
361/361 [==============================] - 1s 2ms/step


 63%|██████▎   | 38/60 [01:31<00:51,  2.34s/it]


Val accuracy (after pruning channel 20 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 65%|██████▌   | 39/60 [01:34<00:49,  2.35s/it]


Val accuracy (after pruning channel 19 ): 98.6056984498138
361/361 [==============================] - 1s 3ms/step


 67%|██████▋   | 40/60 [01:36<00:48,  2.43s/it]


Val accuracy (after pruning channel 43 ): 98.57105741751104
361/361 [==============================] - 1s 2ms/step


 68%|██████▊   | 41/60 [01:39<00:45,  2.41s/it]


Val accuracy (after pruning channel 58 ): 98.53641638520828
361/361 [==============================] - 1s 2ms/step


 70%|███████   | 42/60 [01:41<00:43,  2.40s/it]


Val accuracy (after pruning channel 3 ): 98.19000606218066
361/361 [==============================] - 1s 3ms/step


 72%|███████▏  | 43/60 [01:43<00:40,  2.36s/it]


Val accuracy (after pruning channel 42 ): 97.65307006148784
361/361 [==============================] - 1s 2ms/step


 73%|███████▎  | 44/60 [01:46<00:36,  2.30s/it]


Val accuracy (after pruning channel 1 ): 97.50584567420108
361/361 [==============================] - 1s 3ms/step


 73%|███████▎  | 44/60 [01:48<00:39,  2.47s/it]


Val accuracy (after pruning channel 29 ): 95.75647354291158


Clean test accuracy of Refined Model:  95.74434918160561
ASR of Refined Model:  100.0


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [ ]:
# Prune the model with a threshold of 4 and save the repaired model
repaired_model_threshold_4 = apply_pruning_defense(4)
repaired_model_threshold_4.save("model_repaired_threshold_4.h5")

361/361 [==============================] - 4s 3ms/step
Original accuracy on clean data:  98.64899974019225
361/361 [==============================] - 1s 2ms/step
Attack Success Rate (ASR) of BadNet:  100.0
361/361 [==============================] - 1s 2ms/step


  0%|          | 0/60 [00:00<?, ?it/s]

361/361 [==============================] - 1s 2ms/step


  2%|▏         | 1/60 [00:02<02:24,  2.45s/it]


Val accuracy (after pruning channel 0 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  3%|▎         | 2/60 [00:04<02:10,  2.25s/it]


Val accuracy (after pruning channel 26 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


  5%|▌         | 3/60 [00:07<02:17,  2.41s/it]


Val accuracy (after pruning channel 27 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


  7%|▋         | 4/60 [00:09<02:23,  2.56s/it]


Val accuracy (after pruning channel 30 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  8%|▊         | 5/60 [00:12<02:11,  2.39s/it]


Val accuracy (after pruning channel 31 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 10%|█         | 6/60 [00:14<02:08,  2.38s/it]


Val accuracy (after pruning channel 33 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 12%|█▏        | 7/60 [00:16<02:01,  2.29s/it]


Val accuracy (after pruning channel 34 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 13%|█▎        | 8/60 [00:20<02:22,  2.73s/it]


Val accuracy (after pruning channel 36 ): 98.64899974019225
361/361 [==============================] - 2s 4ms/step


 15%|█▌        | 9/60 [00:24<02:39,  3.13s/it]


Val accuracy (after pruning channel 37 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 17%|█▋        | 10/60 [00:26<02:25,  2.90s/it]


Val accuracy (after pruning channel 38 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 18%|█▊        | 11/60 [00:28<02:10,  2.66s/it]


Val accuracy (after pruning channel 25 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 20%|██        | 12/60 [00:30<01:59,  2.48s/it]


Val accuracy (after pruning channel 39 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 22%|██▏       | 13/60 [00:33<01:55,  2.45s/it]


Val accuracy (after pruning channel 41 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 23%|██▎       | 14/60 [00:36<02:11,  2.86s/it]


Val accuracy (after pruning channel 44 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 25%|██▌       | 15/60 [00:39<02:06,  2.82s/it]


Val accuracy (after pruning channel 45 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 27%|██▋       | 16/60 [00:41<01:54,  2.61s/it]


Val accuracy (after pruning channel 47 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 28%|██▊       | 17/60 [00:43<01:45,  2.46s/it]


Val accuracy (after pruning channel 48 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 30%|███       | 18/60 [00:46<01:38,  2.35s/it]


Val accuracy (after pruning channel 49 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 32%|███▏      | 19/60 [00:48<01:33,  2.28s/it]


Val accuracy (after pruning channel 50 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 33%|███▎      | 20/60 [00:51<01:39,  2.49s/it]


Val accuracy (after pruning channel 53 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 35%|███▌      | 21/60 [00:53<01:38,  2.52s/it]


Val accuracy (after pruning channel 55 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 37%|███▋      | 22/60 [00:55<01:31,  2.40s/it]


Val accuracy (after pruning channel 40 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 38%|███▊      | 23/60 [00:58<01:28,  2.39s/it]


Val accuracy (after pruning channel 24 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 40%|████      | 24/60 [01:00<01:25,  2.38s/it]


Val accuracy (after pruning channel 59 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 42%|████▏     | 25/60 [01:02<01:23,  2.38s/it]


Val accuracy (after pruning channel 9 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 43%|████▎     | 26/60 [01:06<01:38,  2.89s/it]


Val accuracy (after pruning channel 2 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 45%|████▌     | 27/60 [01:09<01:31,  2.77s/it]


Val accuracy (after pruning channel 12 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 47%|████▋     | 28/60 [01:11<01:22,  2.57s/it]


Val accuracy (after pruning channel 13 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 48%|████▊     | 29/60 [01:13<01:15,  2.43s/it]


Val accuracy (after pruning channel 17 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 50%|█████     | 30/60 [01:15<01:09,  2.33s/it]


Val accuracy (after pruning channel 14 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 52%|█████▏    | 31/60 [01:18<01:07,  2.33s/it]


Val accuracy (after pruning channel 15 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 53%|█████▎    | 32/60 [01:21<01:11,  2.54s/it]


Val accuracy (after pruning channel 23 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 55%|█████▌    | 33/60 [01:23<01:08,  2.53s/it]


Val accuracy (after pruning channel 6 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 57%|█████▋    | 34/60 [01:26<01:04,  2.48s/it]


Val accuracy (after pruning channel 51 ): 98.64033948211657
361/361 [==============================] - 1s 2ms/step


 58%|█████▊    | 35/60 [01:28<01:01,  2.45s/it]


Val accuracy (after pruning channel 32 ): 98.64033948211657
361/361 [==============================] - 1s 2ms/step


 60%|██████    | 36/60 [01:31<01:00,  2.52s/it]


Val accuracy (after pruning channel 22 ): 98.63167922404088
361/361 [==============================] - 1s 3ms/step


 62%|██████▏   | 37/60 [01:33<00:56,  2.47s/it]


Val accuracy (after pruning channel 21 ): 98.65765999826795
361/361 [==============================] - 1s 4ms/step


 63%|██████▎   | 38/60 [01:37<01:05,  2.98s/it]


Val accuracy (after pruning channel 20 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 65%|██████▌   | 39/60 [01:39<00:57,  2.73s/it]


Val accuracy (after pruning channel 19 ): 98.6056984498138
361/361 [==============================] - 1s 2ms/step


 67%|██████▋   | 40/60 [01:41<00:50,  2.54s/it]


Val accuracy (after pruning channel 43 ): 98.57105741751104
361/361 [==============================] - 1s 2ms/step


 68%|██████▊   | 41/60 [01:43<00:45,  2.41s/it]


Val accuracy (after pruning channel 58 ): 98.53641638520828
361/361 [==============================] - 1s 2ms/step


 70%|███████   | 42/60 [01:46<00:41,  2.33s/it]


Val accuracy (after pruning channel 3 ): 98.19000606218066
361/361 [==============================] - 1s 3ms/step


 72%|███████▏  | 43/60 [01:49<00:46,  2.75s/it]


Val accuracy (after pruning channel 42 ): 97.65307006148784
361/361 [==============================] - 1s 3ms/step


 73%|███████▎  | 44/60 [01:52<00:44,  2.75s/it]


Val accuracy (after pruning channel 1 ): 97.50584567420108
361/361 [==============================] - 1s 2ms/step


 75%|███████▌  | 45/60 [01:54<00:38,  2.57s/it]


Val accuracy (after pruning channel 29 ): 95.75647354291158
361/361 [==============================] - 1s 2ms/step


 77%|███████▋  | 46/60 [01:56<00:34,  2.43s/it]


Val accuracy (after pruning channel 16 ): 95.20221702606739
361/361 [==============================] - 1s 2ms/step


 78%|███████▊  | 47/60 [01:59<00:31,  2.41s/it]


Val accuracy (after pruning channel 56 ): 94.7172425738287
361/361 [==============================] - 1s 2ms/step


 78%|███████▊  | 47/60 [02:01<00:33,  2.58s/it]


Val accuracy (after pruning channel 46 ): 92.09318437689443


Clean test accuracy of Refined Model:  92.1278254091972
ASR of Refined Model:  99.9913397419243


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [ ]:
# Prune the model with a threshold of 10 and save the repaired model
repaired_model_threshold_10 = apply_pruning_defense(10)
repaired_model_threshold_10.save("model_repaired_threshold_10.h5")

361/361 [==============================] - 9s 3ms/step
Original accuracy on clean data:  98.64899974019225
361/361 [==============================] - 1s 2ms/step
Attack Success Rate (ASR) of BadNet:  100.0
361/361 [==============================] - 1s 2ms/step


  0%|          | 0/60 [00:00<?, ?it/s]

361/361 [==============================] - 1s 3ms/step


  2%|▏         | 1/60 [00:02<02:54,  2.95s/it]


Val accuracy (after pruning channel 0 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  3%|▎         | 2/60 [00:05<02:37,  2.72s/it]


Val accuracy (after pruning channel 26 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  5%|▌         | 3/60 [00:07<02:26,  2.56s/it]


Val accuracy (after pruning channel 27 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  7%|▋         | 4/60 [00:10<02:22,  2.54s/it]


Val accuracy (after pruning channel 30 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


  8%|▊         | 5/60 [00:12<02:17,  2.50s/it]


Val accuracy (after pruning channel 31 ): 98.64899974019225
361/361 [==============================] - 1s 4ms/step


 10%|█         | 6/60 [00:15<02:23,  2.67s/it]


Val accuracy (after pruning channel 33 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 12%|█▏        | 7/60 [00:20<03:03,  3.46s/it]


Val accuracy (after pruning channel 34 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 13%|█▎        | 8/60 [00:23<02:43,  3.14s/it]


Val accuracy (after pruning channel 36 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 15%|█▌        | 9/60 [00:25<02:28,  2.92s/it]


Val accuracy (after pruning channel 37 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 17%|█▋        | 10/60 [00:27<02:14,  2.68s/it]


Val accuracy (after pruning channel 38 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 18%|█▊        | 11/60 [00:30<02:09,  2.64s/it]


Val accuracy (after pruning channel 25 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 20%|██        | 12/60 [00:33<02:10,  2.71s/it]


Val accuracy (after pruning channel 39 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 22%|██▏       | 13/60 [00:35<02:03,  2.63s/it]


Val accuracy (after pruning channel 41 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 23%|██▎       | 14/60 [00:38<01:59,  2.59s/it]


Val accuracy (after pruning channel 44 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 25%|██▌       | 15/60 [00:40<01:51,  2.47s/it]


Val accuracy (after pruning channel 45 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 27%|██▋       | 16/60 [00:42<01:47,  2.45s/it]


Val accuracy (after pruning channel 47 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 28%|██▊       | 17/60 [00:45<01:46,  2.47s/it]


Val accuracy (after pruning channel 48 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 30%|███       | 18/60 [00:49<02:03,  2.94s/it]


Val accuracy (after pruning channel 49 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 32%|███▏      | 19/60 [00:52<01:58,  2.90s/it]


Val accuracy (after pruning channel 50 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 33%|███▎      | 20/60 [00:54<01:48,  2.71s/it]


Val accuracy (after pruning channel 53 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 35%|███▌      | 21/60 [00:56<01:39,  2.55s/it]


Val accuracy (after pruning channel 55 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 37%|███▋      | 22/60 [00:59<01:35,  2.51s/it]


Val accuracy (after pruning channel 40 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 38%|███▊      | 23/60 [01:01<01:36,  2.60s/it]


Val accuracy (after pruning channel 24 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 40%|████      | 24/60 [01:04<01:32,  2.57s/it]


Val accuracy (after pruning channel 59 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 42%|████▏     | 25/60 [01:06<01:27,  2.51s/it]


Val accuracy (after pruning channel 9 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 43%|████▎     | 26/60 [01:09<01:24,  2.48s/it]


Val accuracy (after pruning channel 2 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 45%|████▌     | 27/60 [01:11<01:18,  2.39s/it]


Val accuracy (after pruning channel 12 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 47%|████▋     | 28/60 [01:13<01:14,  2.33s/it]


Val accuracy (after pruning channel 13 ): 98.64899974019225
361/361 [==============================] - 1s 3ms/step


 48%|████▊     | 29/60 [01:16<01:16,  2.48s/it]


Val accuracy (after pruning channel 17 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 50%|█████     | 30/60 [01:18<01:15,  2.51s/it]


Val accuracy (after pruning channel 14 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 52%|█████▏    | 31/60 [01:21<01:09,  2.38s/it]


Val accuracy (after pruning channel 15 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 53%|█████▎    | 32/60 [01:23<01:07,  2.41s/it]


Val accuracy (after pruning channel 23 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 55%|█████▌    | 33/60 [01:25<01:03,  2.35s/it]


Val accuracy (after pruning channel 6 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 57%|█████▋    | 34/60 [01:28<01:01,  2.36s/it]


Val accuracy (after pruning channel 51 ): 98.64033948211657
361/361 [==============================] - 1s 3ms/step


 58%|█████▊    | 35/60 [01:30<01:02,  2.49s/it]


Val accuracy (after pruning channel 32 ): 98.64033948211657
361/361 [==============================] - 1s 3ms/step


 60%|██████    | 36/60 [01:33<01:01,  2.55s/it]


Val accuracy (after pruning channel 22 ): 98.63167922404088
361/361 [==============================] - 1s 2ms/step


 62%|██████▏   | 37/60 [01:36<00:57,  2.50s/it]


Val accuracy (after pruning channel 21 ): 98.65765999826795
361/361 [==============================] - 1s 3ms/step


 63%|██████▎   | 38/60 [01:38<00:54,  2.48s/it]


Val accuracy (after pruning channel 20 ): 98.64899974019225
361/361 [==============================] - 1s 2ms/step


 65%|██████▌   | 39/60 [01:40<00:51,  2.46s/it]


Val accuracy (after pruning channel 19 ): 98.6056984498138
361/361 [==============================] - 1s 2ms/step


 67%|██████▋   | 40/60 [01:42<00:47,  2.36s/it]


Val accuracy (after pruning channel 43 ): 98.57105741751104
361/361 [==============================] - 1s 3ms/step


 68%|██████▊   | 41/60 [01:45<00:47,  2.50s/it]


Val accuracy (after pruning channel 58 ): 98.53641638520828
361/361 [==============================] - 1s 3ms/step


 70%|███████   | 42/60 [01:48<00:46,  2.57s/it]


Val accuracy (after pruning channel 3 ): 98.19000606218066
361/361 [==============================] - 1s 2ms/step


 72%|███████▏  | 43/60 [01:50<00:42,  2.51s/it]


Val accuracy (after pruning channel 42 ): 97.65307006148784
361/361 [==============================] - 1s 3ms/step


 73%|███████▎  | 44/60 [01:53<00:39,  2.49s/it]


Val accuracy (after pruning channel 1 ): 97.50584567420108
361/361 [==============================] - 1s 2ms/step


 75%|███████▌  | 45/60 [01:55<00:36,  2.43s/it]


Val accuracy (after pruning channel 29 ): 95.75647354291158
361/361 [==============================] - 1s 2ms/step


 77%|███████▋  | 46/60 [01:58<00:33,  2.41s/it]


Val accuracy (after pruning channel 16 ): 95.20221702606739
361/361 [==============================] - 1s 3ms/step


 78%|███████▊  | 47/60 [02:00<00:32,  2.54s/it]


Val accuracy (after pruning channel 56 ): 94.7172425738287
361/361 [==============================] - 1s 2ms/step


 80%|████████  | 48/60 [02:03<00:31,  2.60s/it]


Val accuracy (after pruning channel 46 ): 92.09318437689443
361/361 [==============================] - 1s 2ms/step


 82%|████████▏ | 49/60 [02:05<00:27,  2.53s/it]


Val accuracy (after pruning channel 5 ): 91.49562656967177
361/361 [==============================] - 1s 3ms/step


 83%|████████▎ | 50/60 [02:08<00:24,  2.46s/it]


Val accuracy (after pruning channel 8 ): 91.01931237550879
361/361 [==============================] - 1s 2ms/step


 85%|████████▌ | 51/60 [02:10<00:21,  2.37s/it]


Val accuracy (after pruning channel 11 ): 89.17467740538669
361/361 [==============================] - 1s 2ms/step


 85%|████████▌ | 51/60 [02:12<00:23,  2.60s/it]


Val accuracy (after pruning channel 54 ): 84.43751623798389


Clean test accuracy of Refined Model:  84.3335931410756
ASR of Refined Model:  77.015675067117


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


### Compiling all the results

In [ ]:
results_df_1 = pd.DataFrame({
                    "Threshold": [2, 4, 10],
                    "Clean test accuracy of Refined Model": [95.74434918160561, 92.1278254091972, 84.3335931410756],
                    "Attack Success Rate": [100.0, 99.9913397419243, 77.015675067117]
                })

results_df_1

Threshold  Clean test accuracy of Refined Model  Attack Success Rate
0          2                             95.744349           100.000000
1          4                             92.127825            99.991340
2         10                             84.333593            77.015675

### Clearly we observe and improvement in attack success rates, which indicates that as the threshold increases, the model develops better resistance to attacks.

# Obtaining the accuracy on clean test data and the attack success rate (on backdoored test data) as a function of the fraction of channels pruned (X)

In [ ]:
# Function to apply pruning defense and generate statistics
def apply_pruning_and_generate_stats():
    target_layer_index = 6
    OriginalModel = keras.models.load_model(neural_network_model_path)
    PrunedModel = keras.models.load_model(neural_network_model_path)

    # Evaluate the original model on clean and poisoned data
    original_accuracy = np.mean(np.equal(np.argmax(OriginalModel.predict(clean_images_test), axis=1), clean_labels_test)) * 100
    print("Original accuracy on clean data: ", original_accuracy)
    original_asr = np.mean(np.equal(np.argmax(OriginalModel.predict(poisoned_images_valid), axis=1), poisoned_labels_valid)) * 100
    print("\nAttack Success Rate (ASR) on poisoned data: ", original_asr)

    # Get intermediate layer activations
    intermediate_activations = extract_intermediate_layer(OriginalModel, target_layer_index).predict(clean_images_valid)
    average_activations = intermediate_activations.mean(axis=(0,1,2))
    sorted_activation_indices = np.argsort(average_activations)
    total_len = len(sorted_activation_indices)

    # Lists to store results for plotting
    clean_data_accuracies = []
    poisoned_data_asrs = []
    pruning_progresses = []

    # Apply pruning iteratively
    pruning_progress = 0
    for channel in sorted_activation_indices:
        # Zeroing out the weights of the target channel
        layer_to_prune = PrunedModel.layers[target_layer_index - 1]
        K.set_value(layer_to_prune.kernel[:, :, :, channel], np.zeros_like(layer_to_prune.kernel[:, :, :, channel]))
        K.set_value(layer_to_prune.bias[channel], np.zeros_like(layer_to_prune.bias[channel]))

        # Create a refined model instance and evaluate
        refined_model = ModifiedModel(OriginalModel, PrunedModel)
        # pruning_progress = (pruning_progress + 1) / len(sorted_activation_indices)
        print(f"\nPruning progress percentage: {(pruning_progress / total_len):.2f}")

        clean_data_accuracy = np.mean(np.equal(np.argmax(refined_model(clean_images_test), axis=1), clean_labels_test)) * 100
        print("Clean data accuracy of Modified Model: ", clean_data_accuracy)

        poisoned_data_asr = np.mean(np.equal(np.argmax(refined_model(poisoned_images_valid), axis=1), poisoned_labels_valid)) * 100
        print("ASR of Refined Model on poisoned data: ", poisoned_data_asr)

        # Store results for plotting
        clean_data_accuracies.append(clean_data_accuracy)
        poisoned_data_asrs.append(poisoned_data_asr)
        pruning_progresses.append(pruning_progress / total_len)
        pruning_progress = pruning_progress + 1

    return pruning_progresses, clean_data_accuracies, poisoned_data_asrs


In [ ]:
# Execute the pruning function and store the results
pruning_progresses, clean_data_accuracies, poisoned_data_asrs = apply_pruning_and_generate_stats()

361/361 [==============================] - 1s 2ms/step
Original accuracy on clean data:  98.64899974019225
361/361 [==============================] - 1s 2ms/step

Attack Success Rate (ASR) on poisoned data:  100.0
361/361 [==============================] - 1s 2ms/step

Pruning progress percentage: 0.00
Clean data accuracy of Modified Model:  98.64899974019225
ASR of Refined Model on poisoned data:  100.0

Pruning progress percentage: 0.02
Clean data accuracy of Modified Model:  98.64899974019225
ASR of Refined Model on poisoned data:  100.0

Pruning progress percentage: 0.03
Clean data accuracy of Modified Model:  98.64899974019225
ASR of Refined Model on poisoned data:  100.0

Pruning progress percentage: 0.05
Clean data accuracy of Modified Model:  98.64899974019225
ASR of Refined Model on poisoned data:  100.0

Pruning progress percentage: 0.07
Clean data accuracy of Modified Model:  98.64899974019225
ASR of Refined Model on poisoned data:  100.0

Pruning progress percentage: 0.08
C

In [ ]:
results_df_2 = pd.DataFrame({
                    "Pruning progress percentage": pruning_progresses,
                    "Clean test accuracy of Modified Model": clean_data_accuracies,
                    "Attack Success Rate (ASR)": poisoned_data_asrs
                })

results_df_2.to_csv("results.csv", index=False)

results_df_2

Pruning progress percentage  Clean test accuracy of Modified Model  \
0                      0.000000                              98.649000   
1                      0.016667                              98.649000   
2                      0.033333                              98.649000   
3                      0.050000                              98.649000   
4                      0.066667                              98.649000   
5                      0.083333                              98.649000   
6                      0.100000                              98.649000   
7                      0.116667                              98.649000   
8                      0.133333                              98.649000   
9                      0.150000                              98.649000   
10                     0.166667                              98.649000   
11                     0.183333                              98.649000   
12                     0.200000                              98.649000   
13                     0.216667                              98.649000   
14                     0.233333                              98.649000   
15                     0.250000                              98.649000   
16                     0.266667                              98.649000   
17                     0.283333                              98.649000   
18                     0.300000                              98.649000   
19                     0.316667                              98.649000   
20                     0.333333                              98.649000   
21                     0.350000                              98.649000   
22                     0.366667                              98.649000   
23                     0.383333                              98.649000   
24                     0.400000                              98.649000   
25                     0.416667                              98.649000   
26                     0.433333                              98.649000   
27                     0.450000                              98.649000   
28                     0.466667                              98.649000   
29                     0.483333                              98.649000   
30                     0.500000                              98.649000   
31                     0.516667                              98.649000   
32                     0.533333                              98.649000   
33                     0.550000                              98.640339   
34                     0.566667                              98.640339   
35                     0.583333                              98.631679   
36                     0.600000                              98.631679   
37                     0.616667                              98.623019   
38                     0.633333                              98.571057   
39                     0.650000                              98.475795   
40                     0.666667                              98.441154   
41                     0.683333                              98.086083   
42                     0.700000                              97.549147   
43                     0.716667                              97.393262   
44                     0.733333                              95.617909   
45                     0.750000                              95.029012   
46                     0.766667                              94.492076   
47                     0.783333                              91.859357   
48                     0.800000                              91.270460   
49                     0.816667                              90.802806   
50                     0.833333                              88.949511   
51                     0.850000                              84.246991   
52                     0.866667                              76.296874   
53      

# Evaluating the Goodnet G models (`ModifiedModel`)

Below are the results for all the 3 goodnet models on the cleaned and poisoned test set.

In [6]:
!python3 CSAW-HackML-2020/lab3/eval.py CSAW-HackML-2020/lab3/data/cl/test.h5 CSAW-HackML-2020/lab3/data/bd/bd_test.h5 model_repaired_threshold_2.h5

2023-12-04 20:50:36.952484: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-12-04 20:50:36.952546: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-12-04 20:50:36.952591: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-12-04 20:50:38.552445: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
401/401 [==============================] - 11s 28ms/step
Clean Classification accuracy: 95.90023382696803
401/401 [==============================] - 10s 25ms/step
Attack Success Rate: 100.0


In [7]:
!python3 CSAW-HackML-2020/lab3/eval.py CSAW-HackML-2020/lab3/data/cl/test.h5 CSAW-HackML-2020/lab3/data/bd/bd_test.h5 model_repaired_threshold_4.h5

2023-12-04 20:51:35.898735: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-12-04 20:51:35.898818: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-12-04 20:51:35.898866: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-12-04 20:51:37.251276: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
401/401 [==============================] - 13s 31ms/step
Clean Classification accuracy: 92.29150428682775
401/401 [==============================] - 9s 23ms/step
Attack Success Rate: 99.98441153546376


In [8]:
!python3 CSAW-HackML-2020/lab3/eval.py CSAW-HackML-2020/lab3/data/cl/test.h5 CSAW-HackML-2020/lab3/data/bd/bd_test.h5 model_repaired_threshold_10.h5

2023-12-04 20:52:14.858570: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-12-04 20:52:14.859074: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-12-04 20:52:14.859148: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-12-04 20:52:16.190355: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
401/401 [==============================] - 11s 28ms/step
Clean Classification accuracy: 84.54403741231489
401/401 [==============================] - 10s 25ms/step
Attack Success Rate: 77.20966484801247


| Model Name                     | Clean Classification Accuracy (%) | Attack Success Rate (%) |
|--------------------------------|-----------------------------------|-------------------------|
| model_repaired_threshold_2.h5  | 95.90                             | 100.00                  |
| model_repaired_threshold_4.h5  | 92.29                             | 99.98                   |
| model_repaired_threshold_10.h5 | 84.54                             | 77.21                   |
